source/extensions/filters/udp/dns_filter/dns_parser.cc

```diff
@@ -210,20 +392,197 @@ DnsQueryRecordPtr DnsMessageParser::parseDnsQueryRecord(const Buffer::InstancePt
   record_type = buffer->peekBEInt<uint16_t>(name_offset);
   name_offset += sizeof(record_type);
 
-  // Read the record class. This value is almost always 1 for internet address records
+  // Read the record class. This value is always 1 for internet address records
   uint16_t record_class;
   record_class = buffer->peekBEInt<uint16_t>(name_offset);
   name_offset += sizeof(record_class);
 
-  auto rec = std::make_unique<DnsQueryRecord>(record_name, record_type, record_class);
+  if (record_class != DNS_RECORD_CLASS_IN) {
+    ENVOY_LOG(debug, "Unsupported record class '{}' in address record", record_class);
+    return nullptr;
+  }
 
   // stop reading he buffer here since we aren't parsing additional records
-  ENVOY_LOG(trace, "Extracted query record. Name: {} type: {} class: {}", rec->name_, rec->type_,
-            rec->class_);
+  ENVOY_LOG(trace, "Extracted query record. Name: {} type: {} class: {}", record_name, record_type,
+            record_class);
 
   *offset = name_offset;
+  return std::make_unique<DnsQueryRecord>(record_name, record_type, record_class);
+}
+
+void DnsMessageParser::setDnsResponseFlags(DnsQueryContextPtr& query_context,
+                                           const uint16_t questions, const uint16_t answers) {
+  // Copy the transaction ID
+  generated_.id = header_.id;
+
+  // Signify that this is a response to a query
+  generated_.flags.qr = 1;
+
+  generated_.flags.opcode = header_.flags.opcode;
+  generated_.flags.aa = 0;
+  generated_.flags.tc = 0;
+
+  // Copy Recursion flags
+  generated_.flags.rd = header_.flags.rd;
+
+  // Set the recursion flag based on whether Envoy is configured to forward queries
+  generated_.flags.ra = recursion_available_;
+
+  // reserved flag is not set
+  generated_.flags.z = 0;
+
+  // Set the authenticated flags to zero
+  generated_.flags.ad = 0;
+
+  generated_.flags.cd = 0;
+  generated_.answers = answers;
+  generated_.flags.rcode = query_context->response_code_;
+
+  // Set the number of questions from the incoming query
+  generated_.questions = questions;
+
+  // We will not include any additional records
+  generated_.authority_rrs = 0;
+  generated_.additional_rrs = 0;
+}
+
+void DnsMessageParser::buildDnsAnswerRecord(DnsQueryContextPtr& context,
+                                            const DnsQueryRecord& query_rec, const uint32_t ttl,
+                                            Network::Address::InstanceConstSharedPtr ipaddr) {
+  // Verify that we have an address matching the query record type
+  switch (query_rec.type_) {
+  case DNS_RECORD_TYPE_AAAA:
+    if (ipaddr->ip()->ipv6() == nullptr) {
+      ENVOY_LOG(debug, "Unable to return IPV6 address for query");
+      return;
+    }
+    break;
+
+  case DNS_RECORD_TYPE_A:
+    if (ipaddr->ip()->ipv4() == nullptr) {
+      ENVOY_LOG(debug, "Unable to return IPV4 address for query");
+      return;
+    }
+    break;
+
```

Comment:
can we add a TODO to support the other records like SRV

Reply:
Sure